substrate/pkg/controller/substrate/internetgateway.go

```diff
@@ -19,115 +19,97 @@ import (
 	})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("creating internet gateway, %w", err)
 	}
-	logging.FromContext(ctx).Infof("Successfully created internet-gateway %v for cluster %v", *output.InternetGateway.InternetGatewayId, identifier)
-	return output.InternetGateway, nil
+	logging.FromContext(ctx).Infof("Created internet gateway %s", substrate.Name)
+	return createInternetGatewayOutput.InternetGateway, nil
 }
 
-func (i *internetGateway) attachInternetGWToVPC(ctx context.Context, igwID, vpcID string) error {
-	if _, err := i.ec2Client.AttachInternetGatewayWithContext(ctx, &ec2.AttachInternetGatewayInput{
-		InternetGatewayId: aws.String(igwID),
-		VpcId:             aws.String(vpcID),
-	}); err != nil {
-		return err
+func (i *internetGateway) Delete(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	describeVpcsOutput, err := i.ec2Client.DescribeVpcsWithContext(ctx, &ec2.DescribeVpcsInput{Filters: filtersFor(substrate.Name)})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("describing vpc, %w", err)
 	}
-	logging.FromContext(ctx).Infof("Successfully attached internet-gateway %s to VPC ID %s", igwID, vpcID)
-	return nil
-}
-
-func (i *internetGateway) getInternetGateway(ctx context.Context, identifier string) (*ec2.InternetGateway, error) {
-	output, err := i.ec2Client.DescribeInternetGatewaysWithContext(ctx, &ec2.DescribeInternetGatewaysInput{
-		Filters: ec2FilterFor(identifier),
-	})
+	if len(describeVpcsOutput.Vpcs) == 0 {
+		return reconcile.Result{}, nil
+	}
+	describeInternetGatewaysOutput, err := i.ec2Client.DescribeInternetGatewaysWithContext(ctx, &ec2.DescribeInternetGatewaysInput{Filters: filtersFor(substrate.Name)})
 	if err != nil {
-		return nil, fmt.Errorf("describing internet-gateway, %w", err)
+		return reconcile.Result{}, fmt.Errorf("describing internet gateways, %w", err)
 	}
-	if len(output.InternetGateways) == 0 {
-		return nil, nil
+	if len(describeInternetGatewaysOutput.InternetGateways) == 0 {
+		return reconcile.Result{}, nil
 	}
-	if len(output.InternetGateways) > 1 {
-		return nil, fmt.Errorf("expected to find one internet-gateway, but found %d", len(output.InternetGateways))
+	if _, err := i.ec2Client.DetachInternetGatewayWithContext(ctx, &ec2.DetachInternetGatewayInput{
+		VpcId: describeVpcsOutput.Vpcs[0].VpcId, InternetGatewayId: describeInternetGatewaysOutput.InternetGateways[0].InternetGatewayId,
+	}); err != nil {
+		if err.(awserr.Error).Code() == "DependencyViolation" {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		if err.(awserr.Error).Code() != "Gateway.NotAttached" {
+			return reconcile.Result{}, fmt.Errorf("detaching internet gateway, %w", err)
+		}
+	} else {
+		logging.FromContext(ctx).Infof("Deleted internet gateway %s attachent to %s", aws.StringValue(describeVpcsOutput.Vpcs[0].VpcId), aws.StringValue(describeInternetGatewaysOutput.InternetGateways[0].InternetGatewayId))
```

Comment:
I wanted a common language to avoid readers needing to parse the difference of this misaligned aws verbs. I'm gonna overhaul logging in a separate change.